qchem/pennylane_qchem/qchem/structure.py

```diff
@@ -1051,6 +1051,113 @@ def excitations_to_wires(singles, doubles, wires=None):
     return singles_wires, doubles_wires
 
 
+def derivative(H, x, i, delta=0.005291772):
+    r"""Compute the derivative :math:`\partial \hat{H}(x)/\partial x_i` of the electronic Hamiltonian
+    with respect to the :math:`i`-th nuclear coordinate using a central difference approximation.
+
+    .. math::
+
+        \frac{\partial \hat{H}(x)}{\partial x_i} \approx \frac{\hat{H}(x_i+\delta/2)
+        - \hat{H}(x_i-\delta/2)}{\delta}
+
+    Args:
+        H (callable): function with signature ``H(x)`` that builds the electronic
+            Hamiltonian of the molecule for a given set of nuclear coordinates ``x``
+        x (array[float]): 1D array with the nuclear coordinates given in Angstroms.
+            The size of the array should be ``3*N`` where ``N`` is the number of atoms
+            in the molecule.
+        i (int): index of the nuclear coordinate involved in the derivative
+            :math:`\partial \hat{H}(x)/\partial x_i`
+        delta (float): Step size in Angstroms used to displace the nuclear coordinate.
+            Its default value corresponds to 0.01 Bohr radius.
+
+    Returns:
+        pennylane.Hamiltonian: the derivative of the Hamiltonian
+        :math:`\partial \hat{H}(x)/\partial x_i`
+
+    **Example**
+
+    >>> def H(x):
+    ...     return qml.qchem.molecular_hamiltonian(['H', 'H'], x)[0]
+
+    >>> x = np.array([0., 0., 0.35, 0., 0., -0.35])
+    >>> print(derivative(H, x, 2))
+    (-0.7763135743293005) [I0]
+    + (-0.08534360840293387) [Z0]
+    + (-0.08534360840293387) [Z1]
+    + (0.2669341092545041) [Z2]
+    + (0.26693410925450134) [Z3]
+    + (-0.025233628744274508) [Z0 Z1]
+    + (0.0072162443961340415) [Y0 X1 X2 Y3]
+    + (-0.0072162443961340415) [Y0 Y1 X2 X3]
+    + (-0.0072162443961340415) [X0 X1 Y2 Y3]
+    + (0.0072162443961340415) [X0 Y1 Y2 X3]
+    + (-0.030654287745411964) [Z0 Z2]
+    + (-0.023438043349280003) [Z0 Z3]
+    + (-0.023438043349280003) [Z1 Z2]
+    + (-0.030654287745411964) [Z1 Z3]
+    + (-0.02494407786332001) [Z2 Z3]
+    """
+
+    to_bohr = 1.8897261254535
+
+    # plus
+    x_plus = x.copy()
+    x_plus[i] += delta * 0.5
+
+    # minus
+    x_minus = x.copy()
+    x_minus[i] -= delta * 0.5
+
+    return (H(x_plus) - H(x_minus)) * (delta * to_bohr) ** -1
+
+
+def gradient(H, x, delta=0.005291772):
```

Comment:
_gradient_ is a proper name for this function but it might be confused with the gradient of energy in the context of geometry optimisation.